### ISSUE TITLE
Android: Cannot create asynchronous query while in a write transaction

### ISSUE THREAD

**Issue by cristianoccazinsp** [realm/realm-js#3889]
## Goals

Listen to a collection.

## Expected Results

Collection listener to succeed

## Actual Results

Error while trying to listen to a collection: `Cannot create asynchronous query while in a write transaction
Cannot create asynchronous query while in a write transaction`

## Code Sample

The log happens as follows. I'm adding the helper functions for clarity as well.

```
export function asyncQuery(realm, cb) {
  return new Promise((resolve, reject) => {
    try {
      if (!realm.isInTransaction) {
        resolve(cb(realm));
      } else {
        //console.warn("Realm (query) was already in a transaction, delaying call one more loop.");
        setTimeout(() => {
          try {
            resolve(cb(realm));
          } catch (err) {
            reject(err);
          }
        });
      }
    } catch (err) {
      reject(err);
    }
  });
}

export async function getQueueCollection() {
  return await asyncQuery(global.realm, (realm) => {
    return realm.objects('QueueItem');
  });
}

try {
  // CRASH HERE:
  this.queueListener = await getQueueCollection();
  this.queueListener.addListener(this.handleRealmUpdate);
} catch (err) {
  console.error(
    'CRITICAL Queue: Failed to listen to queue collection: ',
    err.message || err,
  );
  Alert.alert(
    'Error',
    'There was an unexpected error and the app needs to be restarted.',
  );
  throw err;
}
```

Issue started happening somewhere between 10.5.x and 10.6.1 (did not happen in 10.4.x).

## Version of Realm and Tooling

- Realm JS SDK Version: 10.6.1
- Node or React Native: React Native 0.62.2
- Client OS & Version: Android 11 / Pixel 5
- Which debugger for React Native: None / Production

**Comment by takameyer** [comment]
@cristianoccazinsp Thanks for reporting and providing some sample code.  We will start investigating this.

**Comment by cristianoccazinsp** [comment]
Might be worth also checking https://github.com/realm/realm-js/issues/3114 as these two errors are happening together in our app out of a sudden.

Adding more info. We keep a global realm instance available throughout the entire app (something like `global.realm = await Realm.open({...})'`.  This global instance is used from the moment the app is mounted until the process dies (we never call `realm.close`).  When the android app starts in headless mode, it is initialized the same way, and if the app is mounted afterwards, the global instance is re-used (never re-initialized).

Lastly, one would say, just catch the error and create a new realm instance. However, we also keep components mounted and listening to some collection changes while the app is mounted. So if we were to close and re-open the realm, all those components listening to collection changes would stop working as they have no way to detect that they have to re create their listeners.

In short, we have 1 or two globally mounted components (persistent while the app is alive) listening to a globally-created realm instance. This global realm instance is also used for every write and query operation, and we never create any additional realm instances nor destroy existing ones.

**Comment by takameyer** [comment]
Glancing at the core code, this error is thrown when a listener is added within a `write` transaction.

**Comment by cristianoccazinsp** [comment]
The code does not add a listener within a write call, but looking at the other error, it looks like the realm is "stuck" in the transaction state. So anything that checks if there's a current transaction returns true and pretty much everything fails.

**Comment by kneth** [comment]
@cristianoccazinsp Does it only happen after a while (12 hours)?

> UPDATE: just realized that this user is also using Android 12 (still in beta?)

I don't think we have tested on Android 12 yet. We did have a change related to Android 12 (emulator only?) in v10.5.0.

**Comment by cristianoccazinsp** [comment]
@kneth this happened at some point in time 5 minutes after the app was started, and then continued to happen permanently for 12 hours straight (from the last log above). The user's logs are something like this:

- 5m before the first error, the app was started in background (headless) and RealmJS was initialized, no errors here.
- The user 5 minutes later opens the app (it was already running due to the above, so the realm is not re-opened as it is already in opened and stored in a global variable), the entire React tree starts to mount. The two components that mount globally and listen to collection changes (2 different collections) fail to mount as `Cannot create asynchronous query while in a write transaction` is thrown when adding the listener.
- The user closes the app, but it stays running in background (cached process)
- 3 hours later, our headless task kicks in again (it runs every 3 hours to do some background uploads), the realm is not re-opened as it is still stored in the global variable (the app never died). All Realm write operation fail with the same error at this point.
- 3 hours later, headless task kicks in again, same issue as above.
- 3 hours later... and so on, until the app process finally died.

In short, the app was opened automatically in background (headless), realm was opened and stored in a global variable. Shortly after the user opens the app and the realm instance that was opened before ends up stuck and unable to perform write operations for the next 12 hours.

**Comment by elle-j** [comment]
@cristianoccazinsp, just to update you, once [this PR](https://github.com/realm/realm-js/pull/5801/) is merged and v11.9.0 is released, the issue you were experiencing may have been resolved.

Thanks for the detailed information!
